src/packets.rs

```diff
@@ -164,9 +164,109 @@ impl<'a> Iterator for PacketListIterator<'a> {
-const PACKET_SIZE: usize = 8 +      // MIDIPacket::timeStamp: MIDITimeStamp/UInt64
-                           2;       // MIDIPacket::length: UInt16
+const PACKET_LIST_HEADER_SIZE: usize = 4;  // MIDIPacketList::numPackets: UInt32
+const PACKET_HEADER_SIZE: usize = 8 +      // MIDIPacket::timeStamp: MIDITimeStamp/UInt64
+                                  2;       // MIDIPacket::length: UInt16
+
+const INLINE_PACKET_BUFFER_SIZE: usize = 28; // must be divisible by 4
+
+enum PacketBufferStorage {
+    // using u32 ensures correct alignment (required on ARM)
+    Inline([u32; INLINE_PACKET_BUFFER_SIZE / 4]),
+    External(Vec<u32>)
+}
+
+impl PacketBufferStorage {
+    #[inline]
+    fn get_slice(&self) -> &[u8] {
+        unsafe {
+            match *self {
+                PacketBufferStorage::Inline(ref inline) =>
+                    slice::from_raw_parts(inline.as_ptr() as *const u8, inline.len() * 4),
+                PacketBufferStorage::External(ref vec) =>
+                    slice::from_raw_parts(vec.as_ptr() as *const u8, vec.len() * 4)
+            }
+        }
+    }
+    
+    #[inline]
+    fn get_slice_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            match *self {
+                PacketBufferStorage::Inline(ref mut inline) =>
+                    slice::from_raw_parts_mut(inline.as_mut_ptr() as *mut u8, inline.len() * 4),
+                PacketBufferStorage::External(ref mut vec) =>
+                    slice::from_raw_parts_mut(vec.as_mut_ptr() as *mut u8, vec.len() * 4)
+            }
+        }
+    }
+    
+    unsafe fn assign_packet(&mut self, offset: usize, time: MIDITimeStamp, data: &[u8]) {
+        assert!(data.len() <= MAX_PACKET_DATA_LENGTH, "packet data too long"); // cannot store longer size in u16
+        if cfg!(any(target_arch = "arm", target_arch = "aarch64")) {
+            debug_assert!(offset & 0b11 == 0);
+        }
+        let slice = self.get_slice_mut();
+        let mut ptr = slice[offset..].as_mut_ptr() as *mut Packet;
+        (*ptr).inner.timestamp = time;
+        (*ptr).inner.length = data.len() as u16;
+        let packet_data_start = offset + PACKET_HEADER_SIZE;
+        slice[packet_data_start..(packet_data_start + data.len())].copy_from_slice(data);
+    }
+
+    /// Requires that there is a valid Packet at `offset`, which has enough space for `data`
+    unsafe fn extend_packet(&mut self, offset: usize, data: &[u8]) {
+        let slice = self.get_slice_mut();
+        let mut ptr = slice[offset..].as_mut_ptr() as *mut Packet;
+        let packet_data_start = offset + PACKET_HEADER_SIZE + (*ptr).inner.length as usize;
+        (*ptr).inner.length += data.len() as u16;
+        slice[packet_data_start..(packet_data_start + data.len())].copy_from_slice(data);
+    }
+    
+    /// Call this only with larger length values (won't make the buffer smaller)
+    unsafe fn set_len(&mut self, new_length: usize) {
+        if new_length < INLINE_PACKET_BUFFER_SIZE { return; }
```

Comment:
If it is an `External` case it could have a length shorter than `INLINE_PACKET_BUFFER_SIZE`, right ?

Reply:
The way I've implemented it, the `External` case will only ever be created with sizes larger than `INLINE_PACKET_BUFFER_SIZE` (see the implementation of `new`).

Reply:
I should probably document this somewhere ...

Reply:
I see.